### DISCUSSION TITLE
Retry design

### DISCUSSION THREAD

**Discussion by souleb** [oras-project/oras-go#232] [Ideas]
Hello All, 

I would like to open the discussion on design HTTP retries as of #147.

# Proposal

Add HTTP Level retry with exponential backoff to the [remote](https://github.com/oras-project/oras-go/tree/main/registry/remote) package.

**Note**: Only GET and HEAD requests can be retried. Retrying PUT/PATCH/POST requests is problematic. 
For uploading the blob, the POST request is retriable but PATCH and PUT are not.

**Note**: The proposed API is based on the [go-retryablehttp](https://github.com/hashicorp/go-retryablehttp) api.

We propose adding 2 new types:

```go
type CheckRetry func(ctx context.Context, resp *http.Response, err error) (bool, error)

type Backoff func(min, max time.Duration, attemptNumber int, resp *http.Response) time.Duration
```

```go
// Client is an auth-decorated HTTP client.
// Its zero value is a usable client that uses http.DefaultClient with no cache.
type Client struct {
	// Client is the underlying HTTP client used to access the remote
	// server.
	// If nil, http.DefaultClient is used.
	Client *http.Client

	// Header contains the custom headers to be added to each request.
	Header http.Header

	// Credential specifies the function for resolving the credential for the
	// given registry (i.e. host:port).
	// `EmptyCredential` is a valid return value and should not be considered as
	// an error.
	// If nil, the credential is always resolved to `EmptyCredential`.
	Credential func(context.Context, string) (Credential, error)

	// Cache caches credentials for direct accessing the remote registry.
	// If nil, no cache is used.
	Cache Cache

	// ClientID used in fetching OAuth2 token as a required field.
	// If empty, a default client ID is used.
	// Reference: https://docs.docker.com/registry/spec/auth/oauth/#getting-a-token
	ClientID string

	// ForceAttemptOAuth2 controls whether to follow OAuth2 with password grant
	// instead the distribution spec when authenticating using username and
	// password.
	// References:
	// - https://docs.docker.com/registry/spec/auth/jwt/
	// - https://docs.docker.com/registry/spec/auth/oauth/
	ForceAttemptOAuth2 bool

  CheckRetry CheckRetry

  Backoff Backoff

  RetryWaitMin time.Duration 
  RetryWaitMax time.Duration
  RetryMax     int           
}
```

`CheckRetry` is called after each request. If it returns false, the Client stops
retrying and returns the response to the caller. If it returns an error, that error
value is returned in lieu of the error from the request. The Client will close any
response body when retrying, but if the retry is aborted it is up to the
CheckRetry callback to properly close any response body before returning.

`Backoff` is called after a failing request to determine the amount of time that
should pass before trying again.

`RetryWaitMin` and `RetryWaitMax` are the minimum and maximum amount of time to wait for a retry,
and are used to determine the backoff time.

`RetryMax` is the maximum number of times to retry.

Users can implement their own `CheckRetry` and `Backoff` functions to customize
the retry behavior.

We propose adding the following default functions to the Client:

```go
func DefaultCheckRetry(ctx context.Context, resp *http.Response, err error) (bool, error) {
  // Don't retry the following errors:
  // too many redirects
  // scheme error
  // TLS handshake error

  // Retry all other errors.

	  // Retry the following status codes:
  // >= 500: Internal Server Error and different from the notImplemented status code.
  // 429: Too Many Requests. This is a rate limit error.
  ...
}

func DefaultBackoff(min, max time.Duration, attemptNumber int, resp *http.Response) time.Duration {
  // Exponential backoff with jitter randomness.
	...
}

func rateLimitBackoff(min, max, resp) time.Duration {
	// Use this when we are rate limited.
	// token Bucket algorithm.
	...
}
```

The `rateLimitBackoff` function is used when we are rate limited and is implemented using
[golang.org/x/time/rate](https://pkg.go.dev/golang.org/x/time/rate). The `RateLimit-Limit`
header is retrieved from the response header and is used to initialize the rate limit bucket.

The DefaultBackoff function is used when we are not rate limited and the proposal is to port
[polly-contrib WaitAndRetry with jitter](https://github.com/Polly-Contrib/Polly.Contrib.WaitAndRetry#wait-and-retry-with-jittered-back-off) code.

## Design considerations

We should be able to retry all HTTP requests.

The response body should be read and closed before retrying, so that we can reuse the connection.

If we rewind the request body before each retry, we should be able to retry the PUT/PATCH/POST request as well.
A good example is the net/http package [rewindBody](https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/net/http/transport.go;drc=c391156f96357593fa18fccee305401e3f82a1a6;l=663).

## Alternatives

Use the [go-retryablehttp](https://github.com/hashicorp/go-retryablehttp) and provide
a custom `CheckRetry` and `Backoff` functions.

**Comment by souleb** [comment]
Any Comments are welcomed! /cc @shizhMSFT @FeynmanZhou

**Comment by Wwwsylvia** [comment]
Hi @souleb, thanks for the design proposal!
I have a question regarding licensing.  Are you proposing that we borrow the design/implementation of [go-retryablehttp](https://github.com/hashicorp/go-retryablehttp)? In this case, how would the license apply?
I found they are using [Mozilla Public License 2.0](https://github.com/hashicorp/go-retryablehttp/blob/master/LICENSE), which states that 
> Permissions of this weak copyleft license are conditioned on making available source code of licensed files and modifications of those files under the same license (or in certain cases, one of the GNU licenses).

**Comment by souleb** [comment]
Regarding [go-retryablehttp](https://github.com/hashicorp/go-retryablehttp) my proposition is to borrow the design, by having the same methods signature for `CheckRetry` and `CheckRetry`. But then we would have a different implementation.

I think a statement referencing [go-retryablehttp](https://github.com/hashicorp/go-retryablehttp) in the source code above those methods shall be enough.

**Comment by shizhMSFT** [comment]
@souleb The [go-retryablehttp](https://github.com/hashicorp/go-retryablehttp) package looks great for HTTP level retry.

Applying the retryable http is also very simple by using [retryClient.StandardClient()](https://github.com/hashicorp/go-retryablehttp#getting-a-stdlib-httpclient-with-retries):
```go
retryClient := retryablehttp.NewClient()
retryClient.RetryMax = 10

authClient := &auth.Client{
    Client: retryClient.StandardClient()
}

// use `authClient` for registry access
```

I don't think we need to re-invent the wheel in `oras-go`. Instead, I'm more interested in the application-level retry.

For example,

```go
type Repository struct {
    // ...

    // Retry specifies a backoff strategy. No retry is applied if nil.
    Retry func(ctx context.Context) BackOff
}

type BackOff interface {
    // Next returns the next backoff duration and whether to stop
    Next() (time.Duration, bool)
}
```

or even integrate with [cenkalti/backoff](https://github.com/cenkalti/backoff), which is in MIT license.

For most operations, we can just retry.
For `Push` related operations, we may further check if `content` is seekable (i.e. implemented `io.Seeker`). If seekable, we can do retry.
For `Fetch` related operations, we may further consider inject retry into the returned `io.ReadCloser`, assuming the remote server support range query.

/cc @qweeah @Wwwsylvia @m5i-work

**Comment by souleb** [comment]
I agree with using [go-retryablehttp](https://github.com/hashicorp/go-retryablehttp). It is in the `Alternatives` section because my understanding was that we wanted to avoid external dependencies.

for the application level retry we could have:

```go
// given an error, a method i.e. Push or Fetch and the content, this function returns a boolean stating whether we can retry
type IsRetryable func(error ErrorCode, method string, content io.Reader) bool
```

When coupling this with a `RetryMax` users could specify when to retry while using a default retry logic based on [cenkalti/backoff](https://github.com/cenkalti/backoff).

**Reply by Wwwsylvia** [reply]
Looks good. `containerd` got something similar: https://github.com/containerd/containerd/blob/main/remotes/docker/resolver.go#L577-L626

**Reply by souleb** [reply]
Great!

Do we want to support the same `errors` for the default behavior?
- http.StatusUnauthorized  => retry here. In case of BearerAuth a refreshToken might be used for example
- http.StatusMethodNotAllowed, Support registries which have not properly implemented the HEAD method for manifests endpoint
- http.StatusTooManyRequests => This is covered by [go-retryablehttp](https://github.com/hashicorp/go-retryablehttp)
- http.StatusRequestTimeout

**Comment by souleb** [comment]
an example based on the discussion so far:
https://gist.github.com/souleb/0b984992a1b705cfa6584c565e7e4420

**Reply by shizhMSFT** [reply]
It seems that everything in this example can be done at the HTTP client level.

**Reply by shizhMSFT** [reply]
Now the question becomes... Should `oras-go` provide a default retryable HTTP client / RoundTripper?

**Comment by shizhMSFT** [comment]
# Retry for ORAS

`oras-go` provides ability to access remote resources including _remote registry servers_ for regular artifacts and _remote blob storage servers_ for foreign layers (see #319). It is worth noting that fetching foreign layers from remote servers is not only specific to remote registry target (i.e. package `registry/remote`) but also applies to other targets like `content/memory`, `content/oci`, and `content/file`.

During the remote access, it is not uncommon to have unstable network connections or temporarily unavailable / faulty cloud servers. Thus a [retry pattern](https://learn.microsoft.com/azure/architecture/patterns/retry) can be applied to mitigate this kind of issues. However, if the unexpected incident of the remote server takes a longer time to be mitigated, an opposite [circuit breaker pattern](https://learn.microsoft.com/azure/architecture/patterns/circuit-breaker) should be applied. Since a suitable pattern should be chosen and applied according to the application specific business logic, `oras-go` should not force applying any pattern by default.

Considering the scenarios of fetching regular layers and foreign layers, it is better to implement retry at the HTTP level instead of the `Repository` or `Registry` structure level. As mentioned by @souleb, [go-retryablehttp](https://github.com/hashicorp/go-retryablehttp) works great with `oras-go` already out of the box. It is not necessarily for `oras-go` to reinvent the wheel.

However, users may not know such packages or they have concerns on the licenses. It might still be good to have a `registry/remote/retry` package in `oras-go` to provide retry ability as an off-the-shelf utility.

## Related Projects

- `github.com/distribution/distribution`: The `registry/client` package does not provide any `retry` logic.
- `github.com/moby/moby`: The `distribution` package implements a retry with exponential back off `0.25s * 2^n` and max `5` attempts.
- `github.com/containerd/containerd`: The `remotes/docker` package implements a retry with no back off and max `6` attempts.
- `github.com/google/go-containerregistry`: The `pkg/v1/remote` package provides a default retry with exponential back off `1s * 3^n ± 10%` and max `3` attempts.

## Proposal

This comment proposes a new package `registry/remote/retry` in `oras-go` to provide a default retry with exponential back off `0.25s * 2^n ± 10%` and max `5` attempts for each request, including auth requests. All HTTP methods should be retried unless the request body is not rewindable. The package MUST NOT cache request body in order to retry.

The retry logic should be implemented to at the `http.RoundTripper` level. Hence, we should have APIs like

```go
package retry

type Transport struct {
    // default to http.DefaultTransport if nil
    Base http.RoundTripper

    // we can export them in future versions of oras-go with better refactoring
    backoff  time.Duration
    factor   int
    attempts int
    jitter   float64
}

var DefaultTransport = &Transport{
    backoff:  250 * time.Millisecond,
    factor:   2,
    attempts: 5,
    jitter:   0.1,
}

var DefaultClient = &http.Client{
    Transport: DefaultTransport,
}

// NewTransport creates an HTTP Transport with the default retry policy
func NewTransport(base http.RoundTripper) *Transport

// NewClient creates an HTTP client with the default retry policy
func NewClient() *http.Client
```

to be developer friendly.

/cc @FeynmanZhou @Wwwsylvia

**Reply by shizhMSFT** [reply]
Retry functionality can be futher decomposed into backoff, predicate, and max attempts for implememtation references.

**Reply by shizhMSFT** [reply]
Since #147 mentioned about `Retry-After`, backoff and predicate can be combined into an adaptive backoff.

**Reply by shizhMSFT** [reply]
@souleb Any inputs are welcome!

**Reply by souleb** [reply]
Thanks for the design.!

What do you think about exporting the backoff parameters? Most users will stick with the default, but some advanced users may want to customize it.

I think the backoff needs to be adaptative based on the `http.Response`.

Regarding the retryable status code, are we following what is stated in https://github.com/oras-project/oras-go/discussions/232#discussioncomment-3503964?

**Reply by shizhMSFT** [reply]
@souleb I tried to follow https://github.com/oras-project/oras-go/discussions/232#discussioncomment-3503964. However, what I end up is exactly the same as the `go-retryablehttp`. For advanced users, why not suggest using `go-retryablehttp` directly?

**Reply by shizhMSFT** [reply]
We can document the hint to use `go-retryablehttp` for advanced users.

**Reply by shizhMSFT** [reply]
Here's what I end up with:
```go
var DefaultClient = &http.Client{
	Transport: &Transport{},
}

var DefaultPolicy Policy = &GenericPolicy{
	Retryable:    DefaultPredicate,
	Backoff:      DefaultBackoff,
	RetryWaitMin: 200 * time.Millisecond, // can be named as MaxWait
	RetryWaitMax: 3 * time.Second,        // can be named as MinWait
	RetryMax:     5,                      // can be named as MaxRetry
}

var DefaultPredicate Predicate = func(ctx context.Context, resp *http.Response, err error) (bool, error) {
	// not implemented
}

var DefaultBackoff Backoff = func(attempt int, resp *http.Response) time.Duration {
	// check Retry-After

	// do linear backoff
	backoff := 250 * time.Millisecond
	factor := 2
	jitter := 0.1
	// not implemented
}

type Policy interface {
	Retry(ctx context.Context, attempt int, resp *http.Response, err error) (time.Duration, error)
}

type Predicate func(ctx context.Context, resp *http.Response, err error) (bool, error)
type Backoff func(attempt int, resp *http.Response) time.Duration

type GenericPolicy struct {
	Retryable    Predicate
	Backoff      Backoff
	RetryWaitMin time.Duration
	RetryWaitMax time.Duration
	RetryMax     int
}

func (p *GenericPolicy) Retry(ctx context.Context, attempt int, resp *http.Response, err error) (time.Duration, error) {
	if attempt > p.RetryMax {
		return -1, err
	}
	if ok, err := p.Retryable(ctx, resp, err); !ok {
		return -1, err
	}
	backoff := p.Backoff(attempt, resp)
	if backoff < p.RetryWaitMin {
		backoff = p.RetryWaitMin
	}
	if backoff > p.RetryWaitMax {
		backoff = p.RetryWaitMax
	}
	return backoff, nil
}

type Transport struct {
	// default to http.DefaultTransport if nil
	Base http.RoundTripper

	// detailed policy implementation
	// factory pattern to allow the policy to store states per request
	Policy func() Policy
}

func (t *Transport) RoundTrip(req *http.Request) (*http.Response, error) {
	ctx := req.Context()
	policy := t.Policy()
	attempt := 0
	for {
		attempt++
		resp, respErr := t.Base.RoundTrip(req)
		duration, err := policy.Retry(ctx, attempt, resp, respErr)
		if duration < 0 {
			return resp, err
		}
		if respErr == nil {
			resp.Body.Close()
		}
		timer := time.NewTimer(duration)
		select {
		case <-ctx.Done():
			timer.Stop()
			return nil, ctx.Err()
		case <-timer.C:
		}
	}
}

// NewTransport creates an HTTP Transport with the default retry policy
func NewTransport(base http.RoundTripper) *Transport {
	return &Transport{
		Base: base,
	}
}

// NewClient creates an HTTP client with the default retry policy
func NewClient() *http.Client {
	return &http.Client{
		Transport: NewTransport(nil),
	}
}
```

**Reply by souleb** [reply]
Hello @shizhMSFT, happy new year.

In your last proposal, you stated:
> However, users may not know such packages or they have concerns on the licenses. It might still be good to have a registry/remote/retry package in oras-go to provide retry ability as an off-the-shelf utility.

I agree that providing docs on how to hook `go-retryablehttp` is a good solution for advance users. As a `oras-go` user I still would like the sdk to provide a fine tuned default client.

Coupling `go-retryablehttp` with those 
```go
var DefaultPolicy Policy = &GenericPolicy{
	Retryable:    DefaultPredicate,
	Backoff:      DefaultBackoff,
	RetryWaitMin: 200 * time.Millisecond, // can be named as MaxWait
	RetryWaitMax: 3 * time.Second,        // can be named as MinWait
	RetryMax:     5,                      // can be named as MaxRetry
}
```
might be just that.

What about the licensing? I think MPL can be used with Apache license.

**Reply by shizhMSFT** [reply]
@souleb Happy New Year :tada:

> I agree that providing docs on how to hook `go-retryablehttp` is a good solution for advance users. As a oras-go user I still would like the sdk to provide a fine tuned default client.

That's exactly what I am thinking about. In `oras-go`, we provide a fine-tuned client, which is optimized for registry access. Most users should just use that client. I think the tunable `GenericPolicy` is good enough for most users and even for some advanced users. For most advanced users, they can use the `go-retryablehttp` package or even develop their own package.

For licensing, MPL can work with Apache license but not always, depending on the users' scenarios.

Here's the plan / proposal: we will create a package `oras.land/oras-go/v2/registry/remote/retry` and make the `retry.DefaultClient` as the default client of `auth.Client`. The `retry` package could look like this:
```go
// client.go

var DefaultClient = NewClient()

// NewClient creates an HTTP client with the default retry policy
func NewClient() *http.Client {
	return &http.Client{
		Transport: NewTransport(nil),
	}
}

type Transport struct {
	// default to http.DefaultTransport if nil
	Base http.RoundTripper

	// detailed policy implementation
	// factory pattern to allow the policy to store states per request
	Policy func() Policy
}

// NewTransport creates an HTTP Transport with the default retry policy
func NewTransport(base http.RoundTripper) *Transport {
	return &Transport{
		Base: base,
	}
}

func (t *Transport) RoundTrip(req *http.Request) (*http.Response, error) {
	ctx := req.Context()
	policy := t.policy()
	attempt := 0
	for {
		attempt++
		resp, respErr := t.Base.RoundTrip(req)
		duration, err := policy.Retry(ctx, attempt, resp, respErr)
		if duration < 0 {
			return resp, err
		}
		if respErr == nil {
			resp.Body.Close()
		}
		timer := time.NewTimer(duration)
		select {
		case <-ctx.Done():
			timer.Stop()
			return nil, ctx.Err()
		case <-timer.C:
		}
	}
}

func (t *Transport) policy() Policy {
	if t.Policy == nil {
		return DefaultPolicy
	}
	return t.Policy()
}
```
and
```go
// policy.go

var DefaultPolicy Policy = &GenericPolicy{
	Retryable: DefaultPredicate,
	Backoff:   DefaultBackoff,
	MinWait:   200 * time.Millisecond,
	MaxWait:   3 * time.Second,
	MaxRetry:  5,
}

var DefaultPredicate Predicate = func(ctx context.Context, resp *http.Response, err error) (bool, error) {
	panic("not implemented")
}

var DefaultBackoff Backoff = LinearBackoff(250*time.Millisecond, 2, 0.1)

type Policy interface {
	Retry(ctx context.Context, attempt int, resp *http.Response, err error) (time.Duration, error)
}

type Predicate func(ctx context.Context, resp *http.Response, err error) (bool, error)

type Backoff func(attempt int, resp *http.Response) time.Duration

func LinearBackoff(backoff time.Duration, factor int, jitter float64) Backoff {
	return func(attempt int, resp *http.Response) time.Duration {
		// check Retry-After

		// do linear backoff with jitter
		panic("not implemented")
	}
}

type GenericPolicy struct {
	Retryable Predicate
	Backoff   Backoff
	MinWait   time.Duration
	MaxWait   time.Duration
	MaxRetry  int
}

func (p *GenericPolicy) Retry(ctx context.Context, attempt int, resp *http.Response, err error) (time.Duration, error) {
	if attempt > p.MaxRetry {
		return -1, err
	}
	if ok, err := p.Retryable(ctx, resp, err); !ok {
		return -1, err
	}
	backoff := p.Backoff(attempt, resp)
	if backoff < p.MinWait {
		backoff = p.MinWait
	}
	if backoff > p.MaxWait {
		backoff = p.MaxWait
	}
	return backoff, nil
}
```

We also want to include this `retry` package in the `2.0.0` GA release, targeting 1/16/2023, as retry is really important for production projects. Would you like to implement it with unit tests and documentation in the following week? Of course, @Wwwsylvia, @akashsinghal, @FeynmanZhou, and @shizhMSFT can help if you are not available or need any help.

**Reply by souleb** [reply]
That's good to me.

**Comment by akashsinghal** [comment]
I’d like to express support for implementation of this. Project Ratify relies upon ORAS-go for multiple simultaneous GET operations to the registry. https://github.com/deislabs/ratify/issues/502

**Comment by BBirke-Vidispine** [comment]
How to best handle the issue with default [_MaxWait_=3 seconds](https://github.com/oras-project/oras-go/blob/2c0bf91ac1594c648e642418e6b13049d90bad00/registry/remote/retry/policy.go#L37) and _429 TooManyRequests_? 
See also issue: https://github.com/oras-project/oras-go/issues/1210
Retry wait time for 429 will usually be at least a minute, possible multiple minutes (example: request limit of 2000/min). The implementation reads the "Retry-After" response header, if available, and returns it as wait time in _Backoff/[ExponentialBackoff](https://github.com/oras-project/oras-go/blob/2c0bf91ac1594c648e642418e6b13049d90bad00/registry/remote/retry/policy.go#L95)_.
However, this _Backoff_ result is then capped to the 3 seconds _MaxWait_ in _Retry_, allowing no meaningful retries of 429 with defaults.

The original [go-retryablehttp](https://github.com/hashicorp/go-retryablehttp) [_DefaultBackoff_](https://github.com/hashicorp/go-retryablehttp/blob/fd004584a46724fae09e2f21d7c382e15c893f42/client.go#L551) caps the wait time already after calculation, but only for default exponential wait, not for wait time from 429 and "Retry-After" header.

What would be best? Longer default _MaxWait_, like 3 minutes? Or apply _MaxWait_ already in _ExponentialBackoff_, but only to actual, exponential wait, not to 429 Retry-After? Or leave everything as it is?
